### DISCUSSION TITLE
Focus method

### DISCUSSION THREAD

**Discussion by danielo515** [MunifTanjim/nui.nvim#123] [Ideas]
Hello.
I think it will be awesome if we can a method to focus any popup/input. I think it adds value when you have to manage several inputs, or you don't want to immediately focus after opening.
Of course you can always go the nvim route and make a combined use of set_current_window with the popup window id followed by `cmd('startinsert')`, but it is not the best experience, and that's why I think a method like this will add value.

Also, having a onFocus event it will be awesome.

**Comment by MunifTanjim** [comment]
All the components expose `.bufnr` and `.winid` to encourage using `vim.api.nvim_*` functions with them.

So you can write your own convenient functions using them, for example:

```lua
local function focus(component)
  vim.api.nvim_set_current_win(component.winid)
end
```

And then use that as:

```lua
focus(popup)
````

> having a onFocus event it will be awesome

I'm not sure but maybe you can use the `WinEnter` autocmd for it? (check `:help WinEnter`)

---

The goal of `nui.nvim` is to provide the basic building blocks so that the user can built on top of it. I don't want to add a method just to wrap a `vim.api.nvim_*` function call.

**Comment by danielo515** [comment]
The thing is there are some extra (and some of them subtle) requirements in
order to make such a method work. Apart from having to access the window
id, you need to set the buffer into input mode, which requires executing a
cmd that is not that easy to find, and on top of all that, you must run all
that within `vim.schedule` or you will face unexpected behaviour. For all
that I thought an abstraction on top of that will be beneficial.

On Mon, Jun 20, 2022 at 10:38 PM Munif Tanjim ***@***.***>
wrote:

> All the component exposes .bufnr and .winid to encourage using
> vim.api.nvim_* functions with them.
>
> So you can write your own convenient functions using them, for example:
>
> local function focus(component)
>   vim.api.nvim_set_current_win(component.winid)end
>
> And then use that as:
>
> focus(popup)
>
> having a onFocus event it will be awesome
>
> I'm not sure but maybe you can use the WinEnter autocmd for it? (check :help
> WinEnter)
> ------------------------------
>
> The goal of nui.nvim is to provide the basic building blocks so that the
> user can built on top of it. I don't want to add a method just to wrap a
> vim.api.nvim_* function call.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/MunifTanjim/nui.nvim/discussions/123#discussioncomment-2988827>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AARKJWKDUQDB7M2I2HCYL6LVQDJDHANCNFSM5Y22ERJQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
---
https://danielorodriguez.com

**Comment by MunifTanjim** [comment]
> Apart from having to access the window id, you need to set the buffer into input mode

That's thing. In your use-case you need to swtich to insert mode. But somebody else will just want to focus it, not change the mode. I don't want to cover everything in `nui.nvim`.

> which requires executing a cmd that is not that easy to find

Well, finding them is just a `:help nvim_<Tab>` away.

> For all that I thought an abstraction on top of that will be beneficial.

If you're writing a plugin using `nui.nvim` and doing those things multiple times, an abstraction on top of `nui.nvim` would absolutely be beneficial. But those abstraction should be made based on what you need. What you need in a particular use-case might not be what other people need for their use-cases.

---

I hope that makes sense 😃
